libs/vkd3d/state.c

```diff
@@ -4453,10 +4611,60 @@ static HRESULT d3d12_pipeline_state_init_graphics_create_info(struct d3d12_pipel
         if (!(graphics->stage_flags & shader_stages_lut[i].stage))
             continue;
 
+        prev_stage = curr_stage;
+        curr_stage = shader_stages_lut[i].stage;
+
+        memset(&io_input_signature, 0, sizeof(io_input_signature));
+
+        /* Ignore errors when a signature is not present. If a missing signature
+         * leads to compatibility issues, validation will fail later anyway. */
+        if (curr_stage == VK_SHADER_STAGE_TESSELLATION_CONTROL_BIT ||
+                curr_stage == VK_SHADER_STAGE_MESH_BIT_EXT)
+            vkd3d_shader_parse_patch_constant_signature(&dxbc, &pc_output_signature);
+
+        if (curr_stage == VK_SHADER_STAGE_TESSELLATION_EVALUATION_BIT)
+            vkd3d_shader_parse_patch_constant_signature(&dxbc, &pc_input_signature);
+
+        if (curr_stage != VK_SHADER_STAGE_VERTEX_BIT &&
+                curr_stage != VK_SHADER_STAGE_TASK_BIT_EXT &&
+                curr_stage != VK_SHADER_STAGE_MESH_BIT_EXT)
+        {
+            vkd3d_shader_parse_input_signature(&dxbc, &io_input_signature);
+
+            if (state->pipeline_type == VKD3D_PIPELINE_TYPE_MESH_GRAPHICS)
+            {
+                /* Mesh shaders are special since register assignment is based on semantics. */
+                if (!vkd3d_validate_mesh_shader_io_signatures(&io_output_signature, &pc_output_signature, &io_input_signature))
+                {
+                    hr = E_INVALIDARG;
+                    goto fail;
+                }
+            }
+            else
+            {
+                if (!vkd3d_validate_shader_io_signatures(prev_stage,
+                        &io_output_signature, curr_stage, &io_input_signature))
+                {
+                    hr = E_INVALIDARG;
+                    goto fail;
+                }
+            }
+        }
+
+        vkd3d_shader_free_shader_signature(&io_input_signature);
+        vkd3d_shader_free_shader_signature(&io_output_signature);
+
+        memset(&io_output_signature, 0, sizeof(io_output_signature));
+
+        /* Read output signature for validation purposes. */
+        if (shader_stages_lut[i].stage != VK_SHADER_STAGE_FRAGMENT_BIT &&
+                shader_stages_lut[i].stage != VK_SHADER_STAGE_TASK_BIT_EXT)
+            vkd3d_shader_parse_output_signature(&dxbc, &io_output_signature);
+
         switch (shader_stages_lut[i].stage)
         {
             case VK_SHADER_STAGE_VERTEX_BIT:
-                if ((ret = vkd3d_shader_parse_input_signature(&dxbc, &input_signature)) < 0)
+                if ((ret = vkd3d_shader_parse_input_signature(&dxbc, &vs_input_signature)) < 0)
```

Comment:
Couldn't this reuse io_input_signature?

Reply:
We need to keep the VS input signature around to deal with the input layout, not just validation.